### ISSUE TITLE
collapsible content widget

### ISSUE THREAD

**Issue by brendanmurphy** [ucla-oarc-web/WebBlocks#151]
I. Collapsible Content
Provide a model pattern for how to allow content-authors to add and designate areas of the Plone CMS the functionality of Collapsible Content.

The recommended solutions should:
·      Leverage existing frameworks
·      Meet current web best-practice
·      Consider the desktop and mobile context in responsive design
·      Meet AA Standards for Accessibility (progressive enhancement)
·      Be visual design agnostic (allow the designers to style with CSS as needed), providing instructions in the prototype on which class/IDs to reference

Some resource points are provided (below links):

·      Please consider the current issues related to collapsible content
·      Review the solutions presented in Twitter Bootrap and the JQuery Library bundled with Plone 4x for gaps meeting above requirements
·      Provide recommendations and outline describing any gaps
·      Produce reproducible set of recommendations
·      Produce a set of coding patterns/anti-patterns related to Collapsible Content so implementation team knows what to benchmarks to meet
·      Produce a consumable code snippet prototype model for the Plone CMS team to successfully integrate

UCLA Use-Case Examples showing actual content:
http://www.ucla.edu/staff
http://commencement.ucla.edu/schedule.cfm

JQuery:
http://docs.jquery.com/UI/Accordion

JQuery Mobile:
http://jquerymobile.com/demos/1.2.0/docs/content/content-collapsible.html

Accessibility Concerns (outline by JQuery Enhance.JS team):
http://filamentgroup.com/lab/expand_and_collapse_content_accessibly_with_progressive_enhancement_jquery/

Enhance JS (as a model for progressive enhancements):
https://github.com/filamentgroup/enhance
https://github.com/filamentgroup/Southstreet

Twitter Bootstrap’s Implementation:
http://twitter.github.com/bootstrap/javascript.html#collapse
How it’s done in Plone (ask Bernard to vet these sources):
http://www.sixfeetup.com/blog/utilize-available-javascript-in-plone-without-knowing-javascript
http://plone.org/documentation/kb/creating-a-collapsed-heading
http://www.uwosh.edu/ploneprojects/docs/how-tos/create-interior-collapsible-regions

Expand and collapse content accessibly with progressive enhancement, jQuery, and ARIA:
http://filamentgroup.com/lab/expand_and_collapse_content_accessibly_with_progressive_enhancement_jquery/

---

Target-Trigger through CSS
from eric (Bollens, Eric ebollens@ucla.edu)
Here's a presentation on it:
http://ebollens.github.com/mwf-conference-slides/introduction_to_core_coding_in_mwf13.html#/5
The implementation itself so far is in MWF repo:
https://github.com/ucla/mwf/blob/feature/js_interactivity/root/assets/js/standard/interactivity.js
https://github.com/ucla/mwf/blob/feature/js_interactivity/root/assets/js/standard/interactivity/expandable.js

**Comment by ebollens** [comment]
While Twitter Bootstrap and jQuery have both done a good job with some of these plugins to date, they are designed in a one-off state where the syntax can vary heavily between two UI effects: the accordion is defined very differently than the modal, where in reality they're actually very similar.

It was this thought that initially drove MWF towards the target-trigger approach. In the variant implemented as an experimental development branch of MWF 1.3, one can use markup as follows:

``` html
<div class="expandable example">
    <p>This div will expand when Trigger is pressed</p>
</div>
<a href="#" class="trigger t-example">Trigger</a>
```

The goal here was to take advantage of a meta-pattern, namely pressing a trigger to cause an effect on a target, and then constructing a modifier set that could allow you to set effects and define your targets and their triggers. 

However, one of the more cumbersome portions of this implementation was that it overloaded the purpose of CSS classes to define triggers. This sucked for two reasons: (1) performance and (2) verbosity. In order to avoid nailing arbitrary elements, we forced the user to set a `trigger` class on any element that had a `t-` class that should be parsed as a trigger. There's a way around this in HTML 5 though: the `data-` attributes. 

John Resig does a great job of explaining the basics of the `data-` attributes here:
http://ejohn.org/blog/html-5-data-attributes/

Using this, we can define our own semantics that let us define triggers.

Consider the following:

``` html
<a href="#" data-target="ele"><!-- ...  --></a>
<div id="ele" data-effect="driver">
    <!-- ...  -->
</div>
```

The idea here is that when you press the `[data-target]` element, it finds the element `#ele` and invokes `[data-effect]` on it, so in this case it would invoke a handler called `driver`.

However, going a step further, drivers should also likely be able to support this (effect on parent element):

``` html
<a href="#" data-target="ele1"><!-- ...  --></a>
<a href="#" data-target="ele2"><!-- ...  --></a>
<ul data-effect="driver">
    <li id="ele1"><!-- ...  --></li>
    <li id="ele2"><!-- ...  --></li>
</ul>
```

This allows you to create things like next and previous buttons right there in the middle of content.

And, of course, what if a button should toggle multiple affects in multiple places? We should allow classes too:

``` html
<a href="#" data-target-class="ele"><!-- ...  --></a>
<div class="ele" data-effect="driver">
    <!-- ...  -->
</div>
<div class="ele" data-effect="driver">
    <!-- ...  -->
</div>
```

My thought is that then we actually use `data-target-class` and `data-target-id` and we alias `data-target` such that `.ele` maps to `data-target-class` while `ele` maps to `data-target-id` (as well of course as allowing `#ele` to map to `data-target-id`).

Using this model, I would envision an accordion as follows:

``` html
<ul id="ele" data-effect="accordion">
    <li><a href="#" data-target="area1"><!-- ...  --></a></li>
    <li id="area1"><!-- ...  --></li>
    <li><a href="#" data-target="area2"><!-- ...  --></a></li>
    <li id="area2"><!-- ...  --></li>
    <li><a href="#" data-target="area3"><!-- ...  --></a></li>
    <li id="area3"><!-- ...  --></li>
</ul>
```

How would this work?
- On initialization, we note the existence of `[data-effect=accordion]`, which causes us to initialize the accordion driver over this element. This could do any basic preparations that are necessary such as hiding all but the first element. I still need to think about exactly how these sorts of preferences would be specified through markup, but the handler would need some easy way to work with them.
- On initialization, a click listener would be attacked to each of the `[data-target]` elements.
- On click of any of the click-listening `[data-target]` elements, it finds the element(s) referenced. For each referenced element, it searches up the inherit tree for a `data-effect` element. When it finds one, it delegates to that driver the role of actually carrying out the effect.

I'm inclined to use a delegate pattern rather than an observer pattern here. That allows the `[data-target]` handler to mediate who actually receives the dispatch, rather than letting each plugin receive it on their own. However, there will need to be a plugin registration phase so that this is extensible without modifying the core.

I'm also thinking there are a bunch of conveniences we might want to provide like the ability to just target the next sibling via `data-target` as opposed to having to specify the ID's specifically. It might actually work better to force the `data-target` to be specified in jQuery selector language (except we wouldn't allow elements, so short hand for IDs could still leave out the pound sign for ease of use).

Just some thoughts. I'll start working up a prototype over the weekend.

**Comment by ebollens** [comment]
What did I do with my Friday night? Wrote a new engine:
https://github.com/ebollens/efx

Take a look at the readme and let me know your thoughts. The engine and its unit tests are pretty much complete, and either over the weekend or early into next week, I'll see about putting together some actual drivers. I think it provides a consistent way to characterize many of the interactions such as expandable, modal, accordion, etc.

**Comment by loganfranken** [comment]
Nice work @ebollens!

Looks good to me; I think Web Blocks will be a good use case for it and it's going to be a lot neater than a bunch of scattered `data-*`-driven JavaScript logic.

Only possible concern (let me know if this should be posted in the efx repo instead) would be that events appear to be bound directly to the `target` elements. This would be optimal in the majority of cases, but it could be a potential problem if there are a _lot_ of targets, in which case you may want to use `.live()` (http://api.jquery.com/live/) or `.delegate()` (http://api.jquery.com/delegate/) (both of which have been superseded by the overloaded `.on()`: http://api.jquery.com/on/).

Not sure if there's a generic way to handle that, though, and it may not be an actual issue: what's that saying about premature optimization again? :wink:

**Comment by ebollens** [comment]
@loganfranken general concept issues are fine here. Post in the Efx tracker if there's a specific problem related to implementation.

The actual way events are bound...
1. Drivers are registered as an (effect, event) tuple. 
2. Listeners are attached to triggers, not targets.
3. An event occurs on a trigger that it has a listener for.
4. The trigger finds all of its targets.
5. For each target, it finds the _closest_ `data-effect` for an existent (effect, event) tuple.
6. It executes the driver for the (effect, event) tuple, passing in the trigger, target and container.

This mechanism makes it possible to write an effect where you put a `data-effect` on a container element and then, when a trigger targets some child element of the one with the `data-effect`, the effect might also make changes to other things in the container (such as the accordion, which would hide all children except the targeted one).

Regarding `.live()` and `.delegate()`, I actually used one of them in the MWF version, but as I approached this version, there were a couple reasons I didn't use them:
- Clarity - We're already introducing layers of indirection with the targets and their effects containers. Adding `.live()` or `.delegate()` creates another layer of indirection. It also isn't really the semantic way of doing things. We are listening on the trigger, so why not listen on the trigger?
- Multi-execution - If you initialize twice with overlapping selectors, the handler will fire multiple times. For attaching event listeners to targets, this is solved during the initialization step (so only once and not when the user is interacting with the interface), whereas with `.live()` and `.delegate()`, it has to be done when the event actually occurs, which slows down the actual event response (and also has to happen every time the event fires).

**Comment by ebollens** [comment]
As a heads up, I've added a first example driver and a demo for it. It's just a simple toggle, plus a couple state modifiers for starting the element visible and using a slide effect rather than just an appear/disappear.

**Comment by ebollens** [comment]
I've added an accordion driver as well.

If you pull the latest version of the repo, you'll see something interesting in the accordion driver (see the second demo in `demo/accordion.html`): if you take the triggers out of the accordion container, the accordion becomes tabs! I need to think about what the meta-pattern is here to rename it, but I suspect that accordion is actually a subset name of the functionality that it actually provides.

As I've put together these drivers, I will say that I'm very happy to see that this sort of trigger-target paradigm really does fit a lot of interactions well. My inclination is that Efx could easily be expanded to include a huge number of drivers for different interesting sorts of interactions (so far as to even support things like drag-and-drop!).

One of the things I'm really liking about it so far: consistency. The way you write CSS is basically the same regardless of if you're doing hide/show, accordion/tabs, or any of the other interactions that can be modeled with this meta-pattern.

Please let me know thoughts if anyone gets a chance to mess around some more with it :-)

**Comment by ebollens** [comment]
I've decided to leave accordion and tabs separate, mostly because their default behavior is different. In accordion, the thing that makes the most sense is to slide content, whereas in tabs, it makes sense to either fade it or just hide/show it. Besides that, though, the implementations are mostly the same. I've gone ahead and committed a `tabs` driver into the `efx` repo.

I will work on merging this in as another WebBlocks package later today.

**Comment by ebollens** [comment]
I have added the Efx repo to the WebBlocks repository, right now with the following drivers:
- Toggle
- Tabs
- Accordion

It's only 4 KB in size, so I've made it a default package (at least for the time being).

While Efx has documentation in its README, I figure we'll add some WebBlocks documentation for it as well that just explains the most clear and succinct ways to use it (since it's pretty flexible in its own right).

Please submit any feature requests or comments to the Efx issue tracker rather than this one:
https://github.com/ebollens/efx/issues
